qiskit/dagcircuit/dagnode.py

```diff
@@ -251,6 +251,7 @@ class DAGOpNode(DAGNode):
     """Object to represent an Instruction at a node in the DAGCircuit."""
 
     __slots__ = ["op", "qargs", "cargs", "sort_key"]
+    _key_cache = {}
```

Comment:
I'm not a huge fan of us encoding this as static storage on the type object, since keys never get garbage collected within the lifecycle of the interpreter.  That said, since they're just integers, it's not necessarily awful, but I'm just super nervous about append-only memory usage.

Reply:
That's fair, I wasn't super worried about it. But with your suggestion below that will solve this too by moving the cache to be scoped to the dag instance.

Reply:
If we scope it by DAG instance, let's make `DAGCircuit.copy` and `.copy_empty_like` both share the same underlying cache dictionary; we don't even need to copy it, just add new references.